File: artifacts/api-server/src/routes/fiado.ts
import { Router, type IRouter } from "express";
import { db } from "@workspace/db";
import { clientesTable, fiadosTable } from "@workspace/db/schema";
import { eq, ilike, and, desc, sql } from "drizzle-orm";
import { CriarClienteBody, PagarFiadoBody } from "@workspace/api-zod";

const router: IRouter = Router();

router.get("/clientes", async (req, res) => {
  const { q } = req.query as { q?: string };
  const condition = q ? ilike(clientesTable.nome, `%${q}%`) : undefined;
  const clientes = await db
    .select()
    .from(clientesTable)
    .where(condition)
    .orderBy(clientesTable.nome);
  res.json(
    clientes.map((c) => ({
      ...c,
      criado_em: c.criado_em.toISOString(),
    }))
  );
});

router.post("/clientes", async (req, res) => {
  const data = CriarClienteBody.parse(req.body);
  const [cliente] = await db.insert(clientesTable).values(data).returning();
  res.status(201).json({ ...cliente, criado_em: cliente.criado_em.toISOString() });
});

router.put("/clientes/:id", async (req, res) => {
  const id = parseInt(req.params.id);
  const data = CriarClienteBody.parse(req.body);
  const [cliente] = await db
    .update(clientesTable)
    .set(data)
    .where(eq(clientesTable.id, id))
    .returning();
  if (!cliente) return res.status(404).json({ ok: false, message: "Cliente não encontrado" });
  res.json({ ...cliente, criado_em: cliente.criado_em.toISOString() });
});

router.get("/clientes/:id/extrato", async (req, res) => {
  const id = parseInt(req.params.id);
  const [cliente] = await db
    .select()
    .from(clientesTable)
    .where(eq(clientesTable.id, id));
  if (!cliente) return res.status(404).json({ ok: false, message: "Cliente não encontrado" });

  const fiados = await db
    .select()
    .from(fiadosTable)
    .where(eq(fiadosTable.cliente_id, id))
    .orderBy(fiadosTable.criado_em);

  const total_aberto = fiados
    .filter((f) => !f.pago)
    .reduce((acc, f) => acc + f.valor, 0);

  res.json({
    cliente: { ...cliente, criado_em: cliente.criado_em.toISOString() },
    fiados: fiados.map((f) => ({
      ...f,
      criado_em: f.criado_em.toISOString(),
      pago_em: f.pago_em?.toISOString() ?? null,
    })),
    total_aberto,
  });
});

router.post("/clientes/:id/pagar", async (req, res) => {
  const id = parseInt(req.params.id);
  const { valor } = PagarFiadoBody.parse(req.body);

  const fiados = await db
    .select()
    .from(fiadosTable)
    .where(and(eq(fiadosTable.cliente_id, id), eq(fiadosTable.pago, false)))
    .orderBy(fiadosTable.criado_em);

  let restante = valor;
  for (const fiado of fiados) {
    if (restante <= 0) break;
    if (restante >= fiado.valor) {
      await db
        .update(fiadosTable)
        .set({ pago: true, pago_em: new Date() })
        .where(eq(fiadosTable.id, fiado.id));
      restante -= fiado.valor;
    } else {
      const valorRestante = fiado.valor - restante;
      await db
        .update(fiadosTable)
        .set({ valor: valorRestante })
        .where(eq(fiadosTable.id, fiado.id));
      await db.insert(fiadosTable).values({
        cliente_id: id,
        venda_id: fiado.venda_id,
        valor: restante,
        pago: true,
        pago_em: new Date(),
      });
      restante = 0;
    }
  }

  res.json({ ok: true, message: "Pagamento registrado" });
});

router.get("/resumo", async (_req, res) => {
  const resultado = await db
    .select({
      cliente: clientesTable,
      total_aberto: sql<number>`sum(${fiadosTable.valor})`.as("total_aberto"),
    })
    .from(fiadosTable)
    .innerJoin(clientesTable, eq(fiadosTable.cliente_id, clientesTable.id))
    .where(eq(fiadosTable.pago, false))
    .groupBy(clientesTable.id)
    .orderBy(desc(sql`sum(${fiadosTable.valor})`));

  res.json(
    resultado.map(({ cliente, total_aberto }) => ({
      cliente: { ...cliente, criado_em: cliente.criado_em.toISOString() },
      total_aberto,
    }))
  );
});

export default router;

File: artifacts/api-server/routers/fiado.py
"""Rotas de Fiado."""
import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from datetime import datetime
from typing import Optional, List
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import func
from database import get_db
from models import Cliente, Fiado
from schemas import CriarClienteInput, ClienteOut, ExtratoClienteOut, PagarFiadoInput, ResumoFiadoOut

router = APIRouter()


def fmt_cliente(c: Cliente) -> dict:
    return {
        "id": c.id,
        "nome": c.nome,
        "apelido": c.apelido,
        "telefone": c.telefone,
        "cpf": c.cpf,
        "observacao": c.observacao,
        "criado_em": c.criado_em.isoformat() if c.criado_em else None,
    }


@router.get("/clientes", response_model=List[ClienteOut])
def listar_clientes(q: Optional[str] = None, db: Session = Depends(get_db)):
    query = db.query(Cliente)
    if q:
        query = query.filter(
            (Cliente.nome.ilike(f"%{q}%")) |
            (Cliente.apelido.ilike(f"%{q}%")) |
            (Cliente.cpf.ilike(f"%{q}%"))
        )
    return [fmt_cliente(c) for c in query.order_by(Cliente.nome).all()]


@router.post("/clientes", response_model=ClienteOut, status_code=201)
def criar_cliente(data: CriarClienteInput, db: Session = Depends(get_db)):
    cliente = Cliente(**data.model_dump())
    db.add(cliente)
    db.commit()
    db.refresh(cliente)
    return fmt_cliente(cliente)


@router.put("/clientes/{id}", response_model=ClienteOut)
def editar_cliente(id: int, data: CriarClienteInput, db: Session = Depends(get_db)):
    cliente = db.query(Cliente).filter(Cliente.id == id).first()
    if not cliente:
        raise HTTPException(status_code=404, detail="Cliente não encontrado")
    for campo, valor in data.model_dump().items():
        setattr(cliente, campo, valor)
    db.commit()
    db.refresh(cliente)
    return fmt_cliente(cliente)


@router.get("/clientes/{id}/extrato")
def extrato_cliente(id: int, db: Session = Depends(get_db)):
    cliente = db.query(Cliente).filter(Cliente.id == id).first()
    if not cliente:
        raise HTTPException(status_code=404, detail="Cliente não encontrado")

    fiados = db.query(Fiado).filter(Fiado.cliente_id == id).order_by(Fiado.criado_em).all()
    total_aberto = sum(f.valor for f in fiados if not f.pago)

    return {
        "cliente": fmt_cliente(cliente),
        "fiados": [
            {
                "id": f.id,
                "venda_id": f.venda_id,
                "valor": f.valor,
                "pago": f.pago,
                "pago_em": f.pago_em.isoformat() if f.pago_em else None,
                "criado_em": f.criado_em.isoformat(),
            }
            for f in fiados
        ],
        "total_aberto": total_aberto,
    }


@router.post("/clientes/{id}/pagar")
def pagar_fiado(id: int, data: PagarFiadoInput, db: Session = Depends(get_db)):
    fiados = (
        db.query(Fiado)
        .filter(Fiado.cliente_id == id, Fiado.pago == False)
        .order_by(Fiado.criado_em)
        .all()
    )

    restante = data.valor
    for fiado in fiados:
        if restante <= 0:
            break
        if restante >= fiado.valor:
            fiado.pago = True
            fiado.pago_em = datetime.utcnow()
            restante -= fiado.valor
        else:
            valor_restante = fiado.valor - restante
            fiado.valor = valor_restante
            novo = Fiado(
                cliente_id=id,
                venda_id=fiado.venda_id,
                valor=restante,
                pago=True,
                pago_em=datetime.utcnow(),
            )
            db.add(novo)
            restante = 0

    db.commit()
    return {"ok": True, "message": "Pagamento registrado"}


@router.get("/resumo")
def resumo_fiado(db: Session = Depends(get_db)):
    resultado = (
        db.query(Cliente, func.sum(Fiado.valor).label("total_aberto"))
        .join(Fiado, Fiado.cliente_id == Cliente.id)
        .filter(Fiado.pago == False)
        .group_by(Cliente.id)
        .order_by(func.sum(Fiado.valor).desc())
        .all()
    )
    return [
        {"cliente": fmt_cliente(c), "total_aberto": float(total)}
        for c, total in resultado
    ]
